Add function to turn named tuples into array of pairs

BraulioVM opened the pull request:
Addresses #26174

I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature


Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add "tupleToNameValuePairs", a function that turns a named tuple into an array of pairs.


Detailed description / Documentation draft:
### `tupleToNameValuePairs`
Turned a named tuple into an array of pairs. For a `Tuple(a T, b T, ..., c T)` return an `Array(Tuple(String, T))`
in which the `String`s represent the named fields of the tuple and the `T`s are the values associated to those names.

#### Example

```
> CREATE TABLE example(col Tuple(CPU double, Memory double)) ENGINE=Memory();
> INSERT INTO example VALUES (tuple(3.4, 5.6));
> SELECT tupleToNameValuePairs(col) FROM example;
┌─tupleToNameValuePairs(col)─────────┐
│ [('CPU',3.4),('Memory',5.6)] │
└──────────────────────────────┘
```


> By adding documentation, you'll allow users to try your new feature immediately, not when someone else will have time to document it later. Documentation is necessary for all features that affect user experience in any way. You can add brief documentation draft above, or add documentation right into your patch as Markdown files in [docs](https://github.com/ClickHouse/ClickHouse/tree/master/docs) folder.

I do want to add documentation, but would like to agree that this function makes sense (as implemented) before
getting the documentation written down.

I have a question: the function, as implemented, doesn't support tuples in which the types have a common super type.
For example:

```
> SELECT tupleToNameValuePairs((toInt8(3), 3))
> Received exception from server (version 21.9.1):
Code: 43. DB::Exception: Received from localhost:9000. DB::Exception: The argument tuple for function tupleToNameValuePairs must contain just one type: While processing tupleToNameValuePairs((toInt8(3), 3)). (ILLEGAL_TYPE_OF_ARGUMENT
```

It would make sense to make this function cast the different values of the subtypes to the supertype. Does it make sense to
get a first version of this function merged in and later iterate on it? Or should I figure out how to do the casting in this PR?

BraulioVM committed:
add `namedTupleItems` function

I'm trying to address issue #26174. This is just a draft so far, as I
don't think it's necessary to force all value types in the tuple to be
the same. It should be possible to cast all values to a shared
supertype. I also have to add tests

BraulioVM committed:
add some basic tests for namedTupleItems

BraulioVM committed:
clean up and add more tests

kitaisreal commented on src/Functions/namedTupleItems.cpp:
```diff
@@ -0,0 +1,130 @@
+class FunctionNamedTupleItems : public IFunction
+{
+public:
+    static constexpr auto name = "namedTupleItems";
```
After discussion with @alexey-milovidov we decided that better name is tupleToNameValuePairs.

kitaisreal commented on src/Functions/namedTupleItems.cpp:
```diff
@@ -0,0 +1,130 @@
+        const DataTypeTuple * tuple = checkAndGetDataType<DataTypeTuple>(col);
+
+        if (!tuple)
+            throw Exception("First argument for function " + getName() + " must "
```
Please consider to use fmt like syntax.
```
throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "First argument for function {} must be a tuple.", getName());
```

BraulioVM replied:
Ah I didn't know there was such a constructor, that's good.

kitaisreal commented on src/Functions/namedTupleItems.cpp:
```diff
@@ -0,0 +1,130 @@
+            throw Exception("First argument for function " + getName() + " must "
+                            "be a tuple.", ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+        const auto& elementTypes = tuple->getElements();
```
Please update it to our style check.

kitaisreal commented on tests/queries/0_stateless/02007_namedtupleitems.sql:
```diff
@@ -0,0 +1,33 @@
+SELECT namedTupleItems(tuple(1, 2, 3));
```
Remove comments.

kitaisreal commented on tests/queries/0_stateless/02007_namedtupleitems.sql:
```diff
@@ -0,0 +1,33 @@
+SELECT namedTupleItems(tuple(1, [1,2])); -- { serverError 43; }
+SELECT namedTupleItems(tuple(1, 'a')); -- { serverError 43; }
+SELECT namedTupleItems(33); -- { serverError 43; }
+
```
Remove additional new line.

kitaisreal commented on tests/queries/0_stateless/02007_namedtupleitems.sql:
```diff
@@ -0,0 +1,33 @@
+-- [('1',1),('2',2),('3',3)]
+
+DROP TABLE IF EXISTS test02007;
+CREATE TABLE test02007 (
```
Update test name to function name with under line separations `tuple_to_name_value_pairs`.

kitaisreal commented on src/Functions/namedTupleItems.cpp:
```diff
@@ -0,0 +1,130 @@
+namespace
+{
+
+/** Extract element of tuple by constant index or name. The operation is essentially free.
```
Please update documentation.

kitaisreal commented on src/Functions/namedTupleItems.cpp:
```diff
@@ -0,0 +1,130 @@
+
+        const auto& firstElementType = elementTypes[0];
+
+        auto it = std::find_if(
```
If I understand correctly we try to find if there is all types in tuple are equal type. From my perspective better to use std::all_of to be more specific.

BraulioVM replied:
you're completely right

kitaisreal commented on tests/queries/0_stateless/02007_namedtupleitems.sql:
```diff
@@ -0,0 +1,33 @@
+-- Disk	6.6
+
+DROP TABLE IF EXISTS test02007;
+SELECT namedTupleItems(tuple(1, 1.3)); -- { serverError 43; } should it?
```
As I understand this tuple contains two different element types UInt8 and Double so exception should be thrown.

BraulioVM replied:
I was wondering if both values should be casted to the smallest supertype of both types. For example:

```
> SELECT namedTupleItems((toInt8(3), 3))
```

users would maybe expect that both types would be casted to a regular `int`. 

If you don't think this is needed, or if you think this could be done in a subsequent PR, I'm perfectly fine with that.

kitaisreal replied:
Current implementation is OK. In a lot of places we require user to be more accurate with types and use additional casts if necessary.

BraulioVM replied:
Sounds good

BraulioVM committed:
change name to function to "tupleToNameValuePairs"

BraulioVM committed:
fixup! clean up and add more tests

kitaisreal commented on src/Functions/tupleToNameValuePairs.cpp:
```diff
@@ -0,0 +1,132 @@
+        const DataTypeTuple * tuple = checkAndGetDataType<DataTypeTuple>(arguments[0].type.get());
+        auto * tuple_col_concrete = assert_cast<const ColumnTuple*>(tuple_col);
+
+        MutableColumnPtr keys = ColumnString::create();
```
Better to put auto here, of specify ColumnString pointer type, it is important because on line 114
```
keys->insertData(key.data(), key.size());
```
can be devirtualized.

BraulioVM replied:
Ha that's a good one

BraulioVM committed:
change type annotation to aid devirtualization

BraulioVM committed:
remove unnecessary comment

The `tupleToNameValuePairs` function is not expected to cast types of
values in a tuple to the least common super type. Users will have to
be explicit about casting in their queries

BraulioVM commented:
I haven't yet run the style checks locally so CI may fail because of them. I will fix any outstanding style problems tomorrow

kitaisreal commented:
@BraulioVM need to fix style check and merge

BraulioVM committed:
remove unused extern declaration

BraulioVM committed:
Merge branch 'master' into add-function/tupleToNameValuePairs

BraulioVM committed:
rename tests

BraulioVM committed:
fixup! rename tests

kitaisreal committed:
Update tupleToNameValuePairs.cpp

gyuton commented:
Internal documentation ticket: DOCSUP-14091.
